### ISSUE TITLE
Facilitating 

### ISSUE THREAD

**Issue by danielbe1** [sschmid/Entitas-CSharp#613]
Hi,

One of the annoyances of having multiple contexts is having logical entities that are decomposed across two or more contexts, two examples that spring to mind are:

- An enemy in an NPC context having a Player character from a PC context as its attack target.
- Player Chracters, NPCs, Projectiles are each a context, and they are represented on the map in a Tiles context

Handling these kinds of relationships require lots of boiler-plate code, and it would be nice if code generation could be utilized to avoid it. It could be nice to have something like:
```
class MetaEntity
{
    public EnemyEntity enemy;
    public PlayerCharacterEntity target;
    public ProjectileEntity hittingProjectile;
}

class MetaSytem<MetaEntity> 
{
    ....

    public void Execute(MetaEntity entity) 
    {
    }

    ...
}
```

Generating the code to create the MetaEntity and MetaSystem would be trivial if the relationship between all three entities would be formalized by indexes or special components. It would really make working with multiple contexts easier and more worthwhile.

WDYT?

**Comment by FNGgames** [comment]
I think you're making your contexts too granular. Players / NPCs / Projectiles (in my mind) are all *Game* entities - they likely share a lot of core state and functionality. My advice is to split your contexts in such a way that you never need this meta-system you're talking about. 

If you really need cross-context connections there are more sensible ways to do this. Instead of this weird MetaEntity with references to three different kinds of entity and it's own special system, why not make it a component in any of those three contexts. F.e. if you have this `ProjectileContext`, you could do

```csharp
[Projectile]
public sealed class ProjectileHitComponent {
    public int projectileId; // the projectile
    public int projectileOwnerId; // who fired it
    public int hitId; // who was hit
}
```

Then on a projectile hit you create a new entity and add this component to it - then you need one reactive system reacting on `ProjectileHit.Added()` and you can get all the entity references and process them as you wish.

```csharp
var bullet = _context.GetEntityWithId(e.projectileHit.projectileId);
var shooter = _context.enemy.GetEntityWithId(e.projectileHit.ownerId);
var victim = _contexts.GetEntityWithId(e.projectileHit.hitId);
```

Honestly I have no idea why you want to have a separate context for a bullet and an enemy and a player.

**Comment by danielbe1** [comment]
That is obviously possible, which would mean I wouldn't be able to enjoy the benefits of having multiple contexts to begin with.

**Comment by FNGgames** [comment]
I guess I'm misunderstanding or missing something, maybe other can help. I've always handled this with components - e.g. CollisionComponent has a handle for the SELF and OTHER entities (plus a third field for collision data). Then i have reactive systems that consume these entities and do stuff. What does the proposed MetaEntity give you that a pure component wouldn't? 

I didn't intend to sound dismissive, sorry if I'm missing your point here.

**Comment by danielbe1** [comment]
Sure you can handle this with pure components, such as a collision event component, but I assume that the collision component is short lived (correct me if I am mistaken) and does not persist beyond the current frame so you don't have to handle cases where one or both of the references become invalid.

But what if you want to represent something long lived like seeking targets? You would either need to perform look ups every frame, or need to write code to make sure references do not become invalid, plus you would probably want to cache results and have all of this tested and verified to work. This can be done by hand but its not so trivial and since it can be done mechanically its best to do it using code generation.

**Comment by bit-chimp** [comment]
I agree with this, being able to use contexts to organise my code better would be incredible.

**Comment by thygrrr** [comment]
I only *kind of* agree.

The only problem is in having an entity that contains a bunch of references to other entities.

I solve this with extensions to entity, which minimizes boilerplate. So if I have an InputEntity e, that has a GameEntity associated with it, through an extension, I can do e.game.<blah>.

It will obviously throw errors (as it should!) if e doesn't have a GameLinkComponent or whatever I call it.
